Add method instrumentation

kjanosz opened the pull request:
Add method instrumentation.
Fix MethodMatcher.Full to allow serialization of it.

kjanosz commented:
I don't check correctness of counting yet. 
I thought that it should be done in actual Measurer tests, which will come in the next PR.

axel22 commented on scalameter-core/src/main/java/org/scalameter/execution/invocation/instrumentation/MethodInvocationCounter.java:
```diff
@@ -0,0 +1,65 @@
+import java.util.concurrent.atomic.AtomicLongArray;
+
+
+/**
```
Please use the following syntax for ScalaDocs:

    /** First line that shortly describes the class.
     *  
     *  More lines that further clarify the meaning of the class.
     *  
     *  '''Note:''' special remarks are highlighted like this.
     */
    public class MethodInvocationCounter {

You should avoid using capital letters inside ScalaDoc comments, as they are not rendered nicely in online docs.

Please apply these guidelines to all your comments.

axel22 commented on scalameter-core/src/main/java/org/scalameter/execution/invocation/instrumentation/MethodInvocationCounter.java:
```diff
@@ -0,0 +1,65 @@
+ *
+ * Stores method invocations in a AtomicLongArray.
+ */
+public class MethodInvocationCounter {
```
Please explain in the comment why you are using the Java class here.

axel22 commented on scalameter-core/src/main/java/org/scalameter/execution/invocation/instrumentation/MethodInvocationCounter.java:
```diff
@@ -0,0 +1,65 @@
+ * Stores method invocations in a AtomicLongArray.
+ */
+public class MethodInvocationCounter {
+    private static volatile boolean _initialized = false;
```
Please add a short explanation for each of these fields.

axel22 commented on scalameter-core/src/main/java/org/scalameter/execution/invocation/instrumentation/MethodInvocationCounter.java:
```diff
@@ -0,0 +1,65 @@
+ */
+public class MethodInvocationCounter {
+    private static volatile boolean _initialized = false;
+    private static volatile boolean _started = false;
```
We do not use underscores in identifier names. For private fields exposed through getters/setters `x`, please use `rawX`.

axel22 commented on scalameter-core/src/main/java/org/scalameter/execution/invocation/instrumentation/MethodInvocationCounter.java:
```diff
@@ -0,0 +1,65 @@
+    /**
+     * Starts measuring of method invocations.
+     */
+    public static synchronized void start() {
```
You should explain who is allowed/intended to call this method and when.

axel22 commented on scalameter-core/src/main/java/org/scalameter/execution/invocation/instrumentation/MethodInvocationCounter.java:
```diff
@@ -0,0 +1,65 @@
+    public static synchronized void start() {
+        if (_initialized && !_started) {
+            _started = true;
+        }
```
Should something else happened if the array has not been initialized? Or already started?

kjanosz replied:
Well, since this method is intended only to be used by invocation counting measurers I thought that it isn't necessary to add any alternative flow.
So maybe the better idea would be to make all this methods (except `methodCalled`) package-restricted and write Scala wrapper inside that package that would call this methods (and it would have appropriate, like `private[scalamater]` access level)?

kjanosz replied:
@axel22: I'm pinging just to be sure that my response wasn't missed.

axel22 replied:
Since the Scala wrapper would also need to be accessible, I'd say that it is fine to leave this as-is.

My comment was more in the sense - why are these methods necessary? Did you want to guard against accidental invocations of `methodCalled` before/after the measurement starts/ends? If so, a comment about this intent would be great.

axel22 commented on scalameter-core/src/main/java/org/scalameter/execution/invocation/instrumentation/MethodInvocationCounter.java:
```diff
@@ -0,0 +1,65 @@
+    /**
+     * Stops measuring of method invocations.
+     */
+    public static synchronized void stop() {
```
Same here.

axel22 commented on scalameter-core/src/main/java/org/scalameter/execution/invocation/instrumentation/MethodInvocationCounter.java:
```diff
@@ -0,0 +1,65 @@
+        if (_initialized && _started) {
+            _started = false;
+        }
+    }
```
If the sequence of states in the `MethodInvocationCounter` is:
- (false, false)
- (true, false)
- (true, true)

kjanosz replied:
Are you suggesting that something should happen on other possible inputs? (your comment looks like it has been trimmed)

axel22 replied:
Sorry, didn't complete it - I wanted to say that if you have the sequence of states: UNINITIALIZED -> INITIALIZED -> STARTED -> STOPPED, maybe you want to model it explicitly like that, instead of using a pair of booleans. It would be less error-prone.

axel22 commented on scalameter-core/src/main/java/org/scalameter/execution/invocation/instrumentation/MethodInvocationCounter.java:
```diff
@@ -0,0 +1,65 @@
+     */
+    public static long[] getCounts() {
+        long[] counts = new long[_counts.length()];
+        for (int i=0; i < _counts.length(); i++) {
```
Please use spaces around `=`.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/InvocationCountMatcher.scala:
```diff
@@ -79,14 +80,18 @@ object InvocationCountMatcher {
     }
 
-    /** Matches method with a [[java.lang.reflect.Method]].
+    /**  Matches method with a [[java.lang.reflect.Method]].
```
Please update this comment in case you no longer wish to use `Method` as ctor param.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/Instrumentation.scala:
```diff
@@ -0,0 +1,107 @@
+import java.io._
+import java.util.jar.{JarEntry, JarOutputStream, Attributes}
+import java.util.zip.ZipFile
+import org.objectweb.asm.{ClassReader, ClassWriter}
```
Please sort the inputs in alphabetic order.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/Instrumentation.scala:
```diff
@@ -0,0 +1,107 @@
+   *  @param in location of classes to read from
+   *  @param p matches if class should be instrumented
+   *  @param instrumentFrom instruments class from a [[java.io.InputStream]]
+   *  @return [[scala.Iterator]] of tuples in a form of (fullClassName, instrumentedClassBytes)
```
Put backticks around this tuple.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/Instrumentation.scala:
```diff
@@ -0,0 +1,107 @@
+   *  Note that it supports reading from directories and jars.
+   *
+   *  @param in location of classes to read from
+   *  @param p matches if class should be instrumented
```
"predicate that decides if a class should be instrumented"

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/Instrumentation.scala:
```diff
@@ -0,0 +1,107 @@
+      def accept(file: File, s: String): Boolean = s.endsWith(".class")
+    }
+
+    def _transformClasses(in: File, rootPath: String): Iterator[(String, Array[Byte])] = {
```
Please avoid the use of underscore.

axel22 replied:
Please add a comment to this method, as it is non-trivial.

kjanosz replied:
As far as I remember there is also an underscore in the two other methods from the previous PRs:
`def _unpickle(times: Int, from: Int)` in the `Pickler.unpickle` and
`def _readBytes()` in the `IO.readFromInputStream`
Should they be also changed? - I can do it in the next PR after we iterate over this.

axel22 replied:
Would be a good idea - I must have missed them during the previous review.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/Instrumentation.scala:
```diff
@@ -0,0 +1,107 @@
+      classes.iterator.collect { case f if p(s"$rootPath/${f.getName}".stripPrefix("/").stripSuffix(".class").replace('/', '.')) =>
+        val path = s"$rootPath/${f.getName}".stripPrefix("/")
+        path -> instrumentFrom(path.stripSuffix(".class"), new FileInputStream(f))
+      } ++ subdirs.iterator.flatMap(f => _transformClasses(f, s"$rootPath/${f.getName}".stripPrefix("/")))
```
Please separate this expression at `++` into two value declarations for readability.
Then call `++` on them as the last step.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/Instrumentation.scala:
```diff
@@ -0,0 +1,107 @@
+    def _transformClasses(in: File, rootPath: String): Iterator[(String, Array[Byte])] = {
+      val subdirs = in.listFiles(dirFilter)
+      val classes = in.listFiles(classFilter)
+      classes.iterator.collect { case f if p(s"$rootPath/${f.getName}".stripPrefix("/").stripSuffix(".class").replace('/', '.')) =>
```
Please break this long line by putting the `case` into the next line indented by two spaces, then indenting the body of the partial function by additional 2 spaces.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/Instrumentation.scala:
```diff
@@ -0,0 +1,107 @@
+   *  @param instrumentFrom instruments class from a [[java.io.InputStream]]
+   *  @return [[scala.Iterator]] of tuples in a form of (fullClassName, instrumentedClassBytes)
+   */
+  private def transformClasses(in: File)(p: String => Boolean)(instrumentFrom: (String, InputStream) => Array[Byte]): Iterator[(String, Array[Byte])] = {
```
Please use the uncurried form.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/Instrumentation.scala:
```diff
@@ -0,0 +1,107 @@
+   *
+   *  Note that it supports reading from directories and jars.
+   *
+   *  @param in location of classes to read from
```
Please add a more detailed doc for these parameters - is the predicate taking an absolute path to the class? Or relative path? Or the full class name?

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/Instrumentation.scala:
```diff
@@ -0,0 +1,107 @@
+   *
+   *  @param in location of classes to read from
+   *  @param p matches if class should be instrumented
+   *  @param instrumentFrom instruments class from a [[java.io.InputStream]]
```
"a function that instruments a given class represented by its classname (?) and the input stream of its contents, and outputs the instrumented class to the resulting byte array"

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/Instrumentation.scala:
```diff
@@ -0,0 +1,107 @@
+    }
+  }
+
+  /** Writes jar with instrumented classes to a given file.
```
with the instrumented classes

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/Instrumentation.scala:
```diff
@@ -0,0 +1,107 @@
+      val zipFile = new ZipFile(in)
+      zipFile.entries().asScala.collect {
+        case entry if p(entry.getName.stripSuffix(".class").replace('/', '.')) =>
+          entry.getName -> instrumentFrom(entry.getName.stripSuffix(".class"), zipFile.getInputStream(entry))
```
Please break this line so that it does not cross the 88 character limit:

    entry.getName -> instrumentFrom(
      entry.getName.stripSuffix(".class"), ...)

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/Instrumentation.scala:
```diff
@@ -0,0 +1,107 @@
+
+  /** Writes jar with instrumented classes to a given file.
+   *
+   *  Note that if given [[org.scalameter.Context]] does not contain classpath
```
if the given

axel22 replied:
does not contain a classpath,

(also add a comma at the end)

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/Instrumentation.scala:
```diff
@@ -0,0 +1,107 @@
+   *  [[org.scalameter.utils.ClassPath.default]] is used to get a default classpath.
+   *
+   *  @param ctx [[org.scalameter.Context]] with a classpath key
+   *  @param matcher [[org.scalameter.execution.invocation.InvocationCountMatcher]] to match methods that needs instrumentation
```
"that need"

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/Instrumentation.scala:
```diff
@@ -0,0 +1,107 @@
+   *  @param to writes jar with instrumented to a given file
+   *  @return lookup table of instrumented methods
+   */
+  def writeInstrumentedClasses(ctx: Context, matcher: InvocationCountMatcher, to: File): Vector[MethodSignature] = {
```
Please use an `ArrayBuffer`, as it is both faster to build, more memory friendly and faster to use for lookups.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/Instrumentation.scala:
```diff
@@ -0,0 +1,107 @@
+    val classpath = ctx.goe(Key.classpath, utils.ClassPath.default)
+    val lookupTable = Vector.newBuilder[MethodSignature]
+    val instrumented: Iterator[(String, Array[Byte])] = classpath.paths.iterator.flatMap { f =>
+      var currentIndex = 0
```
You should either be fully functional and use `zipWithIndex` in the previous line, or you should be imperative, and process all the classes inside a `foreach` loop.

I would advise you to refactor (and rename) the code in the `transformClasses` method to return an `Iterator` over the input classes, instead of pushing the transformation function into that method:

    def filterClasses(in: File)(p: String => Boolean): Iterator[(String, File)] = ...

This gives you several benefits:
- it's easier to debug and understand the code, since the transformation happens in the outer loop instead of in the function object call inside the iterator
- simpler interface for `filterClasses`, which is welcome, since you only use the `filterClasses` once
- better interaction with any mutable state, such as `currentIndex`, which happens as part of the transformation

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/Instrumentation.scala:
```diff
@@ -0,0 +1,107 @@
+      transformClasses(f)(matcher.classMatches) { case (className, classStream)  =>
+        val classfileBuffer = utils.IO.readFromInputStream(classStream)
+        val writer = new ClassWriter(ClassWriter.COMPUTE_MAXS) {
+          override def getCommonSuperClass(type1: String, type2: String): String = {
```
Here you can add a clarifying comment.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/Instrumentation.scala:
```diff
@@ -0,0 +1,107 @@
+        reader.accept(visitor, 0)
+        lookupTable ++= visitor.methods
+        currentIndex = visitor.currentIndex
+        classStream.close()
```
Please do this in a `finally` block.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/Instrumentation.scala:
```diff
@@ -0,0 +1,107 @@
+        writer.toByteArray
+      }
+    }
+
```
Please use one empty line here.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/MethodInvocationCounterVisitor.scala:
```diff
@@ -0,0 +1,45 @@
+import scala.collection.mutable
+
+
+/** Walks over methods in a class and instruments ones that match [[org.scalameter.execution.invocation.InvocationCountMatcher]].
```
"that are matched by..."

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/MethodInvocationCounterVisitor.scala:
```diff
@@ -0,0 +1,45 @@
+ *
+ * @param cv [[org.objectweb.asm.ClassVisitor]] that actually visits classes and methods
+ * @param pattern selects methods that needs to be instrumented
+ * @param counterClass class of counting class
```
Should this be `countedClass`?

kjanosz replied:
No, because it's actually a class that do the counts, which is org.scalameter.execution.invocation.instrumentation.MethodInvocationCounter

axel22 replied:
Sorry - my bad - `counterClass` is correct, obviously.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/MethodInvocationCounterVisitor.scala:
```diff
@@ -0,0 +1,45 @@
+                                                         counterClass: String, counterMethod: String,
+                                                         private[instrumentation] var currentIndex: Int)
+  extends ClassVisitor(Opcodes.ASM5, cv) {
+  private val _methods = mutable.ArrayBuffer.empty[MethodSignature]
```
Please use just `methods` and `className`, unless you plan to expose them as getters/setters, in which case you can use `rawMethods` and `rawClassName` for fields.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/MethodInvocationCounterVisitor.scala:
```diff
@@ -0,0 +1,45 @@
+ * @param pattern selects methods that needs to be instrumented
+ * @param counterClass class of counting class
+ * @param counterMethod method name of counting class
+ * @param currentIndex counting table start index
```
Describe that it is actually a `var` that can be read after the visitor completes its traversal.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/MethodInvocationCounterVisitor.scala:
```diff
@@ -0,0 +1,45 @@
+ */
+private[instrumentation] class MethodInvocationCounterVisitor(cv: ClassVisitor, pattern: InvocationCountMatcher,
+                                                         counterClass: String, counterMethod: String,
+                                                         private[instrumentation] var currentIndex: Int)
```
To make this class more functional, and avoid potential bugs in the future:
1. declare `initialCurrentIndex` here, and mark it as `val`.
2. add a `currentIndex` as a private field.
3. in the `visit` method, which is supposed to be called first, reset both the `currentIndex` and the `methods` buffer.

kjanosz replied:
After a thought, I think that we might not need `var currentIndex`, because it can be computed by using length of a `methods` buffer.

axel22 replied:
Nice observation!

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/MethodSignature.scala:
```diff
@@ -0,0 +1,15 @@
+ *  Note that it does not contain return type.
+ *  {{{
+ *    // method signature for a method `(range: scala.collection.immutable.Range).by(idx: Int)`
+ *    MethodSignature(scala.collection.immutable.Range, "by", "int").toString == scala.collection.immutable.Range.by(int)
```
Is this supposed to contain double quotes?

kjanosz replied:
Ach, my bad. That's right, it should have the following form:
MethodSignature("scala.collection.immutable.Range", "by", "int").toString == "scala.collection.immutable.Range.by(int)"

axel22 commented on scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/MethodSignature.scala:
```diff
@@ -0,0 +1,15 @@
+/** Contains method signature in form of className.methodName(methodArgs...)
+ *  where methodArgs are written as Java types.
+ *
+ *  Note that it does not contain return type.
```
Please add a rationale why.

axel22 commented on scalameter-core/src/test/scala/org/scalameter/execution/invocation/instrumentation/InstrumentationTest.scala:
```diff
@@ -0,0 +1,44 @@
+class InstrumentationTest extends FunSuite with Matchers {
+  class Test {
+    def test1() = "test1"
+    def test2(i: Int, b: String) = ()
```
Please add more methods here, and ensure that the other ones are not matched.

axel22 commented:
Nice work, Chris!

It's fine if you want to add tests in the next PR.

kjanosz commented:
@axel22: Fixed. I've also done a little cleanup, e.g. I've removed overriding of `getCommonSuperClass` since after looking over it's usages in ASM itself it seems to be unnecessary.

axel22 commented on scalameter-core/src/main/java/org/scalameter/execution/invocation/instrumentation/MethodInvocationCounter.java:
```diff
@@ -0,0 +1,74 @@
+ *  Note that this class is a Java class because we want method [[methodCalled]] to be static,
+ *  which makes the bytecode instrumentation easier and more compact.
+ */
+public class MethodInvocationCounter {
```
One more detail - can you use 2 space indentation throughout Java files as well.

axel22 commented:
LGTM

kjanosz commented:
@axel22: Fixed.
